### ISSUE TITLE
Flash attention leads to worse training convergence compared to torch sdpa

### ISSUE THREAD

**Issue by TheTinyTeddy** [Dao-AILab/flash-attention#2151]
During a BF16 training session. I compare the torch sdpa implementation to flash-attention, and found significant training loss convergence difference, where flash-attention leads to a lot higher loss compared to sdpa (and also naive attention which aligns with spda convergence). The q,k,v are BF16 and my system is SM120 with CUDA 13.1

The implementations are

`y = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)`
vs 
`y = flash_attn_func(q, k, v, dropout_p=self.dropout if self.training else 0.0, causal=True)`

Any idea to increase the precision or fix the issue?
